diff --git a/Sample/Platform/DUET/Loader/EfiLdr/X64/Makefile b/Sample/Platform/DUET/Loader/EfiLdr/X64/Makefile
index dc31b13..ba16d9e 100644
--- a/Sample/Platform/DUET/Loader/EfiLdr/X64/Makefile
+++ b/Sample/Platform/DUET/Loader/EfiLdr/X64/Makefile
@@ -20,6 +20,7 @@
 # Globals
 #
 BIN_DIR     = $(BUILD_DIR)\$(PROCESSOR)
+BOOTSECTOR_BIN_DIR = $(EDK_SOURCE)\Sample\Platform\DUET\Loader\BootSector\bin
 TOOLCHAIN   = TOOLCHAIN_$(PROCESSOR)
 
 TOOLBIN_DIR = $(BUILD_DIR)\Tools
@@ -57,15 +58,6 @@ LDRDEP = $(BUILD_DIR)\..\Loader\EfiLdr\Efildr.c \
          $(BUILD_DIR)\..\Loader\EfiLdr\Support.h
 
 all : \
-  $(BIN_DIR)\BootSect.com   \
-  $(BIN_DIR)\Bs16.com       \
-  $(BIN_DIR)\Bs32.com       \
-  $(BIN_DIR)\Gpt.com        \
-  $(BIN_DIR)\Mbr.com        \
-  $(BIN_DIR)\Start64.com    \
-  $(BIN_DIR)\St16_64.com    \
-  $(BIN_DIR)\St32_64.com    \
-  $(BIN_DIR)\Efi64.com2     \
   $(BIN_DIR)\Efildr.efi     \
 
 
@@ -78,72 +70,6 @@ loader :      \
 #
 # Generate loader object
 #
-$(BIN_DIR)\BootSect.obj: $(BUILD_DIR)\..\Loader\BootSector\BootSect.asm
-  $(ASM16) /c /omf /Fo$(BIN_DIR)\BootSect.obj $(BUILD_DIR)\..\Loader\BootSector\BootSect.asm 
-
-$(BIN_DIR)\BootSect.com: $(BIN_DIR)\BootSect.obj
-  cd $(BIN_DIR)
-  $(ASMLINK16) /tiny BootSect.obj,BootSect.com,BootSect.map,,,
-
-$(BIN_DIR)\Bs16.obj: $(BUILD_DIR)\..\Loader\BootSector\Bs16.asm
-  $(ASM16) /c /omf /Fo$(BIN_DIR)\Bs16.obj $(BUILD_DIR)\..\Loader\BootSector\Bs16.asm 
-
-$(BIN_DIR)\Bs16.com: $(BIN_DIR)\Bs16.obj
-  cd $(BIN_DIR)
-  $(ASMLINK16) /tiny Bs16.obj,Bs16.com,Bs16.map,,,
-
-$(BIN_DIR)\Bs32.obj: $(BUILD_DIR)\..\Loader\BootSector\Bs32.asm
-  $(ASM16) /c /omf /Fo$(BIN_DIR)\Bs32.obj $(BUILD_DIR)\..\Loader\BootSector\Bs32.asm 
-
-$(BIN_DIR)\Bs32.com: $(BIN_DIR)\Bs32.obj
-  cd $(BIN_DIR)
-  $(ASMLINK16) /tiny Bs32.obj,Bs32.com,Bs32.map,,,
-
-$(BIN_DIR)\Gpt.obj: $(BUILD_DIR)\..\Loader\BootSector\Gpt.asm
-  $(ASM16) /c /omf /Fo$(BIN_DIR)\Gpt.obj $(BUILD_DIR)\..\Loader\BootSector\Gpt.asm 
-
-$(BIN_DIR)\Gpt.com: $(BIN_DIR)\Gpt.obj
-  cd $(BIN_DIR)
-  $(ASMLINK16) /tiny Gpt.obj,Gpt.com,Gpt.map,,,
-
-$(BIN_DIR)\Mbr.obj: $(BUILD_DIR)\..\Loader\BootSector\Mbr.asm
-  $(ASM16) /c /omf /Fo$(BIN_DIR)\Mbr.obj $(BUILD_DIR)\..\Loader\BootSector\Mbr.asm 
-
-$(BIN_DIR)\Mbr.com: $(BIN_DIR)\Mbr.obj
-  cd $(BIN_DIR)
-  $(ASMLINK16) /tiny Mbr.obj,Mbr.com,Mbr.map,,,
-
-$(BIN_DIR)\Start64.obj: $(BUILD_DIR)\..\Loader\BootSector\Start64.asm
-  $(ASM16) /c /omf /Fo$(BIN_DIR)\Start64.obj $(BUILD_DIR)\..\Loader\BootSector\Start64.asm 
-
-$(BIN_DIR)\Start64.com: $(BIN_DIR)\Start64.obj
-  cd $(BIN_DIR)
-  $(ASMLINK16) /tiny Start64.obj,Start64.com,Start64.map,,,
-
-$(BIN_DIR)\St16_64.obj: $(BUILD_DIR)\..\Loader\BootSector\St16_64.asm
-  $(ASM16) /c /omf /Fo$(BIN_DIR)\St16_64.obj $(BUILD_DIR)\..\Loader\BootSector\St16_64.asm 
-
-$(BIN_DIR)\St16_64.com: $(BIN_DIR)\St16_64.obj
-  cd $(BIN_DIR)
-  $(ASMLINK16) /tiny St16_64.obj,St16_64.com,St16_64.map,,,
-
-$(BIN_DIR)\St32_64.obj: $(BUILD_DIR)\..\Loader\BootSector\St32_64.asm
-  $(ASM16) /c /omf /Fo$(BIN_DIR)\St32_64.obj $(BUILD_DIR)\..\Loader\BootSector\St32_64.asm 
-
-$(BIN_DIR)\St32_64.com: $(BIN_DIR)\St32_64.obj
-  cd $(BIN_DIR)
-  $(ASMLINK16) /tiny St32_64.obj,St32_64.com,St32_64.map,,,
-
-$(BIN_DIR)\Efi64.obj: $(BUILD_DIR)\..\Loader\BootSector\Efi64.asm
-  $(ASM16) /c /omf /Fo$(BIN_DIR)\Efi64.obj $(BUILD_DIR)\..\Loader\BootSector\Efi64.asm 
-
-$(BIN_DIR)\Efi64.com: $(BIN_DIR)\Efi64.obj
-  cd $(BIN_DIR)
-  $(ASMLINK16) /tiny Efi64.obj,Efi64.com,Efi64.map,,,
-
-$(BIN_DIR)\Efi64.com2: $(BIN_DIR)\Efi64.com
-  $(TOOLBIN_DIR)\Splitfile $(BIN_DIR)\Efi64.com 135168
-
 $(BIN_DIR)\Efildr.obj: $(LDRDEP)
   $(CC) $(C_FLAGS) $(BUILD_DIR)\..\Loader\EfiLdr\Efildr.c
 
@@ -172,15 +98,15 @@ Fv\DxeIpl.z : $(BIN_DIR)\DxeIpl.efi
 Fv\Efildr64: $(BIN_DIR)\Efildr.efi Fv\DxeIpl.z Fv\DxeMain.z Fv\EfiMain.z
   $(TOOLBIN_DIR)\Efildrimage Fv\Efildr64 $(BIN_DIR)\Efildr.efi Fv\DxeIpl.z Fv\DxeMain.z Fv\EfiMain.z
 
-Fv\Efildr: $(BIN_DIR)\Start64.com $(BIN_DIR)\Efi64.com2 Fv\Efildr64
-  copy /b $(BIN_DIR)\Start64.com+$(BIN_DIR)\Efi64.com2+Fv\Efildr64 Fv\EfildrPure
+Fv\Efildr: Fv\Efildr64
+  copy /b $(BOOTSECTOR_BIN_DIR)\Start64.com+$(BOOTSECTOR_BIN_DIR)\Efi64.com2+Fv\Efildr64 Fv\EfildrPure
   $(TOOLBIN_DIR)\GenPage Fv\EfildrPure Fv\Efildr
 
-Fv\Efildr16: $(BIN_DIR)\St16_64.com $(BIN_DIR)\Efi64.com2 Fv\Efildr64
-  copy /b $(BIN_DIR)\St16_64.com+$(BIN_DIR)\Efi64.com2+Fv\Efildr64 Fv\Efildr16Pure
+Fv\Efildr16: Fv\Efildr64
+  copy /b $(BOOTSECTOR_BIN_DIR)\St16_64.com+$(BOOTSECTOR_BIN_DIR)\Efi64.com2+Fv\Efildr64 Fv\Efildr16Pure
   $(TOOLBIN_DIR)\GenPage Fv\Efildr16Pure Fv\Efildr16
 
-Fv\Efildr20: $(BIN_DIR)\St32_64.com $(BIN_DIR)\Efi64.com2 Fv\Efildr64
-  copy /b $(BIN_DIR)\St32_64.com+$(BIN_DIR)\Efi64.com2+Fv\Efildr64 Fv\Efildr20Pure
+Fv\Efildr20: Fv\Efildr64
+  copy /b $(BOOTSECTOR_BIN_DIR)\St32_64.com+$(BOOTSECTOR_BIN_DIR)\Efi64.com2+Fv\Efildr64 Fv\Efildr20Pure
   $(TOOLBIN_DIR)\GenPage Fv\Efildr20Pure Fv\Efildr20
 

